blockstack_client/backend/drivers/ipfs.py

```diff
@@ -0,0 +1,410 @@
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+    You should have received a copy of the GNU General Public License
+    along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import sys
+import os
+import io
+import re
+import tempfile
+import zlib
+import time
+import json
+import logging
+import ipfsapi
+from common import get_logger, DEBUG, compress_chunk, decompress_chunk
+from ConfigParser import SafeConfigParser
+
+log = get_logger("blockstack-storage-skel")
+log.setLevel( logging.DEBUG if DEBUG else logging.INFO )
+
+IPFS_SERVER   = 'localhost'
+IPFS_PORT     = '5001'
+IPFS_COMPRESS = True
+
+def ipfs_key_gen(key):
+    ''' 
+    We need a custom wraper for this API endpoint, because the ipfsapi does not provide it, yet
+    Instead of checking whether a key exists before creating it,
+    We create it, and ignore the error that throws if it exists already.
+    '''
+    api = ipfsapi.connect('127.0.0.1', 5001)
+    try:
+      r = api._client.request('/key/gen', (key,), decoder='json', opts={'type':'rsa','size':'2048'})
+    except ipfsapi.exceptions.ErrorResponse:
+      # An exception is thrown when the key already exists, we ignore it
+      pass
+
+def write_chunk( chunk_path, chunk_buf, is_mutable ):
+    """
+    Write a chunk of data to IPFS.
+    
+    Return True on success 
+    Return False on error, and log an exception
+    """
+    
+    api = ipfsapi.connect(IPFS_SERVER, IPFS_PORT)
+
+    rc = True
+    begin = None
+    end = None
+    size = None
+    try:
+        if IPFS_COMPRESS:            
+            compressed_data = compress_chunk( chunk_buf )
+        else:
+            compressed_data = chunk_buf
+
+        size = len(compressed_data)
+
+        begin = time.time()
+        h = api.add_str( compressed_data, decoder='json' )
+        if is_mutable:
+          if os.environ.get("BLOCKSTACK_TEST") == "1":
+            log.debug("[DEBUG] ipfs.publish_name %s: %s" % (chunk_path, chunk_buf))
+          r = api.name_publish( h, opts = {'key': 'mutable-{}'.format(chunk_path.replace( "/", r"\x2f" ))})
+          if os.environ.get("BLOCKSTACK_TEST") == "1":
+            log.debug("[DEBUG] ipfs.publish_name returned " % (r))
+        end = time.time()
+        
+    except Exception, e:
+        log.error("Failed to write '%s'" % chunk_path)
+        log.exception(e)
+        rc = False
+
+    if os.environ.get("BLOCKSTACK_TEST") == "1" and rc:
+        log.debug("[BENCHMARK] s3.write_chunk %s: %s" % (size, end - begin))
+    return rc
+
+#-------------------------
+def read_chunk( chunk_path, is_mutable ):
+    """
+    Get a chunk of data from IPFS.
+    
+    Return the data on success
+    Return None on error, and log an exception.
+    """
+    
+    api = ipfsapi.connect(IPFS_SERVER, IPFS_PORT)
+
+    data = None
+    begin = None
+    end = None
+    size = None
+    try:
+        begin = time.time()
```

Comment:
Yes, I used the s3.py as a base template, I felt it was the easiest to understand and most relevant to implement this other driver. I will remove these lines.